src/cell_list.h

```diff
@@ -0,0 +1,226 @@
+            }
+        }
+        const cell_list<T, CELL_SIZE, MAXX, MAXY> *list;
+        const point origin;
+        const point origin_cell; // = origin / CELL_SIZE
+        const int max_distance;
+        std::vector<std::vector<std::pair<point, T>>> sorted_elements;
+        // Upper left and lower right corners of area, used to determine loaded distance to scan.
+        point min_bounds;
+        point max_bounds;
+        int distance_to_nearest_cell_wall = 0;
+        int cell_ranks_loaded = 0;
+};
+
+/*
+ * cell_list_iterator is a forward iterator over the underlying cell_list_range adaptor.
+ * It iterates from nearest entries to most distant entries, and when it reaches
+ * the specified range limit it halts.
+ * The intended use case is to be invoked in a iterator-based for or while loop with
+ * early termination when a valid match is found.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY>
+class cell_list_iterator
+{
+    public:
+        // The default constructor creates an end iterator.
+        cell_list_iterator() : range( nullptr ) {};
+        cell_list_iterator( cell_list_range< T, CELL_SIZE, MAXX, MAXY > *range ) : range( range ) {
+            current = range->sorted_elements[current_distance].begin();
+            if( range->sorted_elements[current_distance].empty() ) {
+                ++( *this );
+            }
+        }
+        std::pair<point, T> &operator*() const {
+            assert( range );
+            return *current;
+        }
+        std::pair<point, T> *operator->() const {
+            assert( range );
+            return &*current;
+        }
+
+        cell_list_iterator &operator++() {
+            assert( range );
+            if( current != range->sorted_elements[current_distance].end() ) {
+                ++current;
+            }
+            // Effectively, "while current does not point to a valid entry".
+            while( current == range->sorted_elements[current_distance].end() ) {
+                ++current_distance;
+                if( current_distance > range->max_distance ) {
+                    range = nullptr;
+                    break;
+                }
+                if( current_distance > range->distance_to_nearest_cell_wall ) {
+                    range->load();
+                }
+                current = range->sorted_elements[current_distance].begin();
+            }
+            return *this;
+        }
+
+        bool operator==( const cell_list_iterator &rhs ) const {
+            if( range == nullptr && rhs.range == nullptr ) {
+                return true;
+            }
+            if( range != rhs.range ) {
+                return false;
+            }
+            return current == rhs.current;
+        }
+        bool operator!=( const cell_list_iterator &rhs ) const {
+            return !( *this == rhs );
+        }
+
+    private:
+        cell_list_range<T, CELL_SIZE, MAXX, MAXY> *range;
+        int current_distance = 1;
```

Comment:
`sorted_elements` starts with zero, but it's never visited. Is the discrepancy intentional?

Reply:
Just making indexing easier, there's already a ton of index and bounds checking going on here without risking one more off by one error to save a single empty vector.